File: src/android/src/com/che/zap/device/AbstractZapDevice.kt
package com.che.zap.device

import com.che.zap.crypto.EncryptionUtils
import com.che.zap.crypto.LocalKeyProvider
import com.che.zap.crypto.ZapCrypto
import com.che.zap.utils.Logger
import com.che.zap.utils.startsWith
import com.che.zap.utils.toHexString
import timber.log.Timber

abstract class AbstractZapDevice {

    companion object {
        internal const val LOG_RAW = true
    }

    private var devicePublicKeyBytes: ByteArray? = null
    private var localKeyProvider = LocalKeyProvider()
    protected var zapEncryption = ZapCrypto(localKeyProvider)

    fun processCharacteristic(characteristicName: String, bytes: ByteArray?) {
        if (bytes == null) return

        if (LOG_RAW) Timber.d("$characteristicName ${bytes.toHexString()}")

        when {
            bytes.startsWith(ZapConstants.RIDE_ON.plus(ZapConstants.RESPONSE_START)) -> processDevicePublicKeyResponse(bytes)
            bytes.size > Int.SIZE_BYTES + EncryptionUtils.MAC_LENGTH -> processEncryptedData(bytes)
            else -> Logger.e("Unprocessed - Data Type: ${bytes.toHexString()}")
        }
    }

    abstract fun processEncryptedData(bytes: ByteArray)

    fun buildHandshakeStart(): ByteArray {
        return ZapConstants.RIDE_ON.plus(ZapConstants.REQUEST_START).plus(localKeyProvider.getPublicKeyBytes())
    }

    private fun processDevicePublicKeyResponse(bytes: ByteArray) {
        devicePublicKeyBytes = bytes.copyOfRange(ZapConstants.RIDE_ON.size + ZapConstants.RESPONSE_START.size, bytes.size)
        zapEncryption.initialise(devicePublicKeyBytes!!)
        if (LOG_RAW) Logger.d("Device Public Key - ${devicePublicKeyBytes!!.toHexString()}")
    }

}

File: src/ios/zwift_play/abstractZapDevice.swift
import Foundation
import os.log

// Assuming Logger and extensions for Data to convert to hex string and to check prefix are defined elsewhere in your Swift project.
extension Data {
    func toHexString() -> String {
        self.map { String(format: "%02x", $0) }.joined()
    }

    func starts(with prefix: Data) -> Bool {
        self.prefix(prefix.count) == prefix
    }
}

class AbstractZapDevice {
    private static let logRaw = true

    private var devicePublicKeyBytes: Data?
    private let localKeyProvider = LocalKeyProvider()
    open var zapEncryption: ZapCrypto

    init() {
        self.zapEncryption = ZapCrypto(localKeyProvider: localKeyProvider)
    }

    func processCharacteristic(characteristicName: String, bytes: Data?) -> Int {
        guard let bytes = bytes else { return 0 }

        if Self.logRaw {
            os_log("%{public}@ %{public}@", log: OSLog.default, type: .debug, characteristicName, bytes.toHexString())
        }

        if bytes.starts(with: ZapConstants.rideOn + ZapConstants.responseStart) {
            processDevicePublicKeyResponse(bytes: bytes)
        } else if bytes.count > MemoryLayout<Int>.size + EncryptionUtils.macLength {
            return processEncryptedData(bytes: bytes)
        } else {
            // Logger equivalent in Swift
            os_log("Unprocessed - Data Type: %{public}@", log: OSLog.default, type: .error, bytes.toHexString())
        }
        return 0
    }

    func buildHandshakeStart() -> Data {
        return ZapConstants.rideOn + ZapConstants.requestStart + localKeyProvider.getPublicKeyBytes()
    }

    private func processDevicePublicKeyResponse(bytes: Data) {
        let startIndex = ZapConstants.rideOn.count + ZapConstants.responseStart.count
        devicePublicKeyBytes = bytes.subdata(in: startIndex..<bytes.count)
        zapEncryption.initialise(devicePublicKeyBytes: devicePublicKeyBytes!)
        if Self.logRaw {
            // Logger equivalent in Swift
            os_log("Device Public Key - %{public}@", log: OSLog.default, type: .debug, devicePublicKeyBytes!.toHexString())
        }
    }

    // Abstract method placeholder - Swift doesn't support abstract classes/methods directly, so this method should be overridden in subclass.
    func processEncryptedData(bytes: Data) -> Int {
        fatalError("This method should be overridden by subclasses")
    }
}
